common/lib/host_list_provider/monitoring/cluster_topology_monitor.ts

```diff
@@ -0,0 +1,517 @@
+  private ignoreNewTopologyRequestsEndTime: number = -1;
+  private ignoreTopologyRequestMs: number = 10000; // 10 seconds.
+
+  // Tracking of the host monitors.
+  private hostMonitors: Map<string, HostMonitor> = new Map();
+  public hostMonitorsWriterClient = null;
+  public hostMonitorsWriterInfo: HostInfo = null;
+  public hostMonitorsReaderClient = null;
+  public hostMonitorsLatestTopology: HostInfo[] = [];
+
+  // Controls for stopping asynchronous monitoring tasks.
+  private stopMonitoring: boolean = false;
+  public hostMonitorsStop: boolean = false;
+  private untrackedPromises: Promise<void>[] = [];
+
+  // Signals to other methods that asynchronous tasks have completed/should be completed.
+  private requestToUpdateTopology: boolean = false;
+
+  constructor(
+    clusterId: string,
+    topologyMap: CacheMap<string, HostInfo[]>,
+    initialHostInfo: HostInfo,
+    props: Map<string, any>,
+    pluginService: PluginService,
+    hostListProvider: MonitoringRdsHostListProvider,
+    refreshRateMs: number,
+    highRefreshRateMs: number
+  ) {
+    this.clusterId = clusterId;
+    this.topologyMap = topologyMap;
+    this.initialHostInfo = initialHostInfo;
+    this._pluginService = pluginService;
+    this._hostListProvider = hostListProvider;
+    this.refreshRateMs = refreshRateMs;
+    this.highRefreshRateMs = highRefreshRateMs;
+
+    this._monitoringProperties = new Map<string, any>(props);
+    for (const [key, val] of props) {
+      if (key.startsWith(ClusterTopologyMonitorImpl.MONITORING_PROPERTY_PREFIX)) {
+        this._monitoringProperties.set(key.substring(ClusterTopologyMonitorImpl.MONITORING_PROPERTY_PREFIX.length), val);
+        this._monitoringProperties.delete(key);
+      }
+    }
+    this.untrackedPromises.push(this.run());
+  }
+
+  get hostListProvider(): MonitoringRdsHostListProvider {
+    return this._hostListProvider;
+  }
+
+  get pluginService(): PluginService {
+    return this._pluginService;
+  }
+
+  get monitoringProperties(): Map<string, any> {
+    return this._monitoringProperties;
+  }
+
+  async close(): Promise<void> {
+    this.stopMonitoring = true;
+    this.hostMonitorsStop = true;
+    this.requestToUpdateTopology = true;
+    await Promise.all(this.untrackedPromises);
+    await this.closeConnection(this.monitoringClient);
+    await this.closeConnection(this.hostMonitorsWriterClient);
+    await this.closeConnection(this.hostMonitorsReaderClient);
+    this.untrackedPromises = [];
+    this.hostMonitors.clear();
+  }
+
+  async forceMonitoringRefresh(shouldVerifyWriter: boolean, timeoutMs: number): Promise<HostInfo[]> {
+    if (this.ignoreNewTopologyRequestsEndTime > 0 && Date.now() < this.ignoreNewTopologyRequestsEndTime) {
+      // Previous failover has just completed, use results without triggering new update.
+      const currentHosts = this.topologyMap.get(this.clusterId);
+      if (currentHosts !== null) {
```

Comment:
Can we add the `ClusterTopologyMonitorImpl.ignoringTopologyRequest` log here as well?